File: rdkit-sys/wrapper/src/descriptors.cc
#include "rust/cxx.h"
#include <GraphMol/Descriptors/MolDescriptors.h>
#include <GraphMol/Descriptors/Property.h>
#include <GraphMol/ROMol.h>


namespace RDKit {
    using Descriptors::Properties;

    double get_exact_molecular_weight(const std::shared_ptr<ROMol> &mol) {
        return RDKit::Descriptors::calcExactMW(*mol);
    }

    std::shared_ptr<Properties> new_properties() {
        return std::shared_ptr<Properties>(new Properties());
    }

    std::unique_ptr<std::vector<std::string>> get_property_names(const std::shared_ptr<Properties> &props) {
        std::vector<std::string> names = props->getPropertyNames();
        std::vector<std::string> *names_heap = new std::vector<std::string>(names);
        return std::unique_ptr<std::vector<std::string>>(names_heap);
    }

    std::unique_ptr<std::vector<double>> compute_properties(const std::shared_ptr<Properties> &props, const std::shared_ptr<ROMol> &mol) {
        std::vector<double> computed = props->computeProperties(*mol);
        auto computed_heap = new std::vector<double>(computed);
        return std::unique_ptr<std::vector<double>>(computed_heap);
    }
}

File: src/descriptors.rs
use std::collections::HashMap;

use cxx::SharedPtr;

use crate::ROMol;

pub struct Properties {
    ptr: SharedPtr<rdkit_sys::descriptors_ffi::Properties>,
}

impl Default for Properties {
    fn default() -> Self {
        Properties::new()
    }
}

pub fn get_exact_molecular_weight(ro_mol: &ROMol) -> f64 {
    return rdkit_sys::descriptors_ffi::get_exact_molecular_weight(&ro_mol.ptr);
}

impl Properties {
    pub fn new() -> Self {
        Properties {
            ptr: rdkit_sys::descriptors_ffi::new_properties(),
        }
    }

    pub fn compute_properties(&self, ro_mol: &ROMol) -> HashMap<String, f64> {
        let names = rdkit_sys::descriptors_ffi::get_property_names(&self.ptr);
        let computed = rdkit_sys::descriptors_ffi::compute_properties(&self.ptr, &ro_mol.ptr);

        assert!(names.len() != 0);
        assert!(computed.len() == names.len());

        names
            .into_iter()
            .zip(computed.as_slice())
            .map(|(k, v)| (k.to_string(), *v))
            .collect()
    }
}
